state/indexer/block/kv/kv.go

```diff
@@ -82,10 +87,138 @@ func (idx *BlockerIndexer) Index(bh types.EventDataNewBlockEvents) error {
 	if err := idx.indexEvents(batch, bh.Events, height); err != nil {
 		return fmt.Errorf("failed to index FinalizeBlock events: %w", err)
 	}
-
 	return batch.WriteSync()
 }
 
+func getKeys(indexer BlockerIndexer) [][]byte {
+	var keys [][]byte
+
+	itr, err := indexer.store.Iterator(nil, nil)
+	if err != nil {
+		panic(err)
+	}
+	for ; itr.Valid(); itr.Next() {
+		keys = append(keys, itr.Key())
+	}
+	return keys
+}
+
+func (idx *BlockerIndexer) Prune(retainHeight int64) (int64, int64, error) {
+	// Returns numPruned, newRetainHeight, err
+	// numPruned: the number of heights pruned or 0 in case of error. E.x. if heights {1, 3, 7} were pruned and there was no error, numPruned == 3
+	// newRetainHeight: new retain height after pruning or lastRetainHeight in case of error
+	// err: error
+
+	lastRetainHeight, err := idx.getLastRetainHeight()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to look up last block indexer retain height: %w", err)
+	}
+	if lastRetainHeight == 0 {
+		lastRetainHeight = 1
+	}
+
+	batch := idx.store.NewBatch()
+	defer func(batch dbm.Batch) {
+		err := batch.Close()
+		if err != nil {
+			idx.log.Error(fmt.Sprintf("Error when closing pruning batch: %v", err))
```

Comment:
If you replaced the batch (at least once) in line 160, won't you be trying to close the same batch twice? If that is the case, I don't think you should be producing an error here... it'll be misleading to the operator.

Same comment in the other deferred function below

Reply:
From what I understand, `batch.Close()` is idempotent, meaning one can call it multiple times. So closing a closed batch would not lead to an error, though I agree that this is far from being clear.

I didn't come up with this approach myself, but rather I took from [here](https://github.com/cometbft/cometbft/blob/6a7dfb40d0302de5cdbca51cc4daede74b6d44c2/store/store.go#L312)

Reply:
I see, in that case, if you're sure that closing a closed batch returns `nil`, I'm OK with this approach.